sql/pgwire/encoding.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Ben Darnell
+
+package pgwire
+
+import (
+	"encoding/binary"
+	"io"
+)
+
+// bufferedReader is implemented by bufio.Reader and bytes.Buffer.
+type bufferedReader interface {
+	io.Reader
+	ReadString(delim byte) (string, error)
+	ReadByte() (byte, error)
+}
+
+// readMsg reads a length-prefixed message. It is only used directly
+// during the authentication phase of the protocol; readTypedMsg is
+// used at all other times.
+func readUntypedMsg(rd io.Reader) ([]byte, error) {
+	var size int32
+	if err := binary.Read(rd, binary.BigEndian, &size); err != nil {
+		return nil, err
+	}
+	// size includes itself.
+	size -= 4
+	data := make([]byte, size)
```

Comment:
We limit the size of keys and values at the KV layer. Perhaps some multiple of that.